frontend/components/music_player/player_hooks.js

```diff
@@ -0,0 +1,261 @@
+  const [timeCounter, setTimeCounter] = useState(0);
+  const [prevVol, setPrevVol] = useState(100);
+  const [volume, setVolume] = useState(100);
+  const [volumeHover, setVolumeHover] = useState(false);
+  const [shuffle, setShuffle] = useState(false);
+  const [repeat, setRepeat] = useState(false);
+  const [songsPlayed, setSongsPlayed] = useState([]);
+  const [duration, setDuration] = useState([]);
+
+  let intervals = [];
+
+  const audioEl = React.createRef();
+  const progressSlider = React.createRef();
+  const volumeSlider = React.createRef();
+
+  // first useEffect is to load songs into the player's queue
+  // cleans up the player by setting universal state to 'pause' when player is closed
+  useEffect(() => {
+    props.getSongs().then((songs) => {
+      props.loadQueue(Object.values(songs.songs));
+    });
+    return () => {
+      props.pauseSong();
+    };
+  }, []);
+
+  // pause and play
+  const handleAction = () => {
+    if (props.songPlaying) {
+      props.pauseSong();
+      audioEl.current.pause();
+    } else {
+      props.playSong();
+      audioEl.current.play();
+    }
+  };
+
+  // if muting, saves the volume into a variable of 'previous volume', sets the volume to zero.
+  // if unmuting, sets the volume of the player to what the volume was before muting
+  const handleMute = () => {
+    if (!props.songMuted) {
+      setPrevVol(audioEl.current.volume * 100);
+      props.muteSong();
+      audioEl.current.volume = 0;
+      setVolume(0);
+    } else if (props.songMuted) {
+      props.muteSong();
+      audioEl.current.volume = prevVol / 100;
+      setVolume(prevVol);
+    }
+  };
+
+  // listener function is used so that when a new song loads up, the music player will know to keep playing music
+  // issue: this is firing off any time any change or interaction is made with the player - how do i get this to run only when song changes?
+  const listener = () => {
+    if (audioEl.current) {
+      if (!props.songPlaying) {
+        audioEl.current.pause();
+      } else {
+        audioEl.current.play();
+      }
+      setDuration(Math.floor(audioEl.current.duration));
+    }
+  };
+  useEffect(listener, [audioEl]);
+
+  // puts the time in the timeCounter or duration to a nice 0:00 format
+  const formatTime = (time) => {
+    if (!time) {
+      time = 0;
+    }
+    let min = Math.floor(time / 60);
+    let sec = Math.floor(time - min * 60);
+    let secstring = sec < 10 ? `0${sec}` : `${sec}`;
+    return `${min}:${secstring}`;
+  };
+
+  // sets up a function that checks the current time of the song every 70ms. will also clear this interval when player closes
+  const timeCounterFunction = () => {
+    if (props.songPlaying) {
+      const interval = setInterval(() => {
+        if (props.songPlaying) {
+          setTimeCounter(Math.floor(audioEl.current.currentTime)), 70;
+        }
+      });
+      intervals.push(interval);
+    }
+    return () => {
+      clearInterval(intervals[0]);
```

Comment:
You shouldn't be clearing things that live outside the scope of the effect. You also shouldn't be pushing to arrays outside the scope of the effect. It should all be enclosed